src/erlfmt_ast.erl

```diff
@@ -0,0 +1,1086 @@
+        {record_name, Pos, Name} ->
+            AAnno = dummy_anno(),
+            erlfmt_to_st_1({tuple, Pos, [{atom, AAnno, '*record_name*'}, Name]});
+        %% A new node `{concat, Anno, Concatables}`, where `Concatables` is a
+        %% list of `string`, `var`, and `macro_call` nodes. This is used to
+        %% represent implicit string concatenation, for example `"foo" "bar"`.
+        {concat, Pos, Subtrees} ->
+            AAnno = dummy_anno(),
+            erlfmt_to_st_1({tuple, Pos, [{atom, AAnno, '*concat*'} | Subtrees]});
+        %% A new node `{macro_string, Anno, Name}` is introduced, where `Name`
+        %% is either an `atom` or a `var` node. It represents `??Name`.
+        {macro_string, Pos, Name} ->
+            AAnno = dummy_anno(),
+            erlfmt_to_st_1({tuple, Pos, [{atom, AAnno, '*stringify*'}, Name]});
+        %% erlfmt preserves '...' tokens as nodes (which erl_parse doesn't)
+        {'...', Pos} ->
+            AAnno = dummy_anno(),
+            erlfmt_to_st_1({tuple, Pos, [{atom, AAnno, '*...*'}]});
+        %% sometimes erlfmt leaves comments as separate nodes
+        %% instead of attaching them to another node
+        {comment, Pos, Lines} ->
+            erl_syntax:set_pos(erl_syntax:comment(Lines), Pos);
+        %% erlfmt has a separate entry for shebang nodes; we use raw strings
+        {shebang, Pos, Text} ->
+            erlfmt_to_st({raw_string, Pos, Text});
+        %% args nodes may (in macros) occur free floating
+        {args, Pos, Args} ->
+            AAnno = dummy_anno(),
+            erlfmt_to_st_1({tuple, Pos, [{atom, AAnno, '*args*'} | Args]});
+        %% TODO:
+        %% New `{spec_clause, Anno, Head, Body, Guards}` node for clauses
+        %% inside `spec` and `callback` attributes, similar to the `clause`
+        %% node above. It reflects the fact that in specs guards come after
+        %% body. The `Head` element is always an `args` node.
+
+        _ ->
+            %% all remaining cases can be handled by the default erl_syntax
+            %% subtree traversal
+            erlfmt_to_st_1(Node)
+    end.
+
+%% assuming erl_parse format, compatible with erl_syntax
+%% TODO: should convert erlfmt anno to erl_syntax pos+annotation
+erlfmt_to_st_1(Node) ->
+    case erl_syntax:subtrees(Node) of
+        [] ->
+            % leaf node
+            Node;
+        Groups0 ->
+            %% recurse and replace the subtrees
+            Groups1 = erlfmt_subtrees_to_st(Groups0),
+            erl_syntax:update_tree(Node, Groups1)
+    end.
+
+erlfmt_subtrees_to_st(Groups) ->
+    [
+        [
+            erlfmt_to_st(Subtree)
+            || Subtree <- Group
+        ]
+        || Group <- Groups
+    ].
+
+get_function_name([{clause, _, {call, _, Name, _}, _, _} | _]) ->
+    %% take the name node of the first clause with a call shape
+    %% TODO: this loses info if not all clauses have the same name
```

Comment:
So is this future work and do we have a missing test case?

Reply:
There is future work to maybe extend erl_syntax to preserve separate name nodes for each clause. Right now, nothing can be done here.